src/librustc_mir/transform/promote_consts.rs

```diff
@@ -741,6 +753,14 @@ pub fn validate_candidates(
             // and `#[rustc_args_required_const]` arguments here.
 
             let is_promotable = validator.validate_candidate(candidate).is_ok();
+
+            if validator.explicit && !is_promotable {
+                ccx.tcx.sess.delay_span_bug(
```

Comment:
Could you add a comment explaining why this check is absolutely important?
I want to be sure someone doesn't just remove it when it seems convenient.^^

The `explicit` field's doc comment already says a few things, so here's a proposal:
```
// If we use explicit validation, we carry the risk of turning a legitimate run-time
// operation into a failing compile-time operation. Make sure that does not happen
// by asserting that there is no possible run-time behavior here in case promotion fails.
```

Reply:
What actually happens is not that we turn it into a failing compile-time op, it's that we don't promote at all, thus keeping the runtime operation, which happened to be the right thing in the cases that the assertion caught.

Reply:
> keeping the runtime operation

We do that *because we don't even try*. But the branch here is for after we tried, so we cannot be in the case where we want to keep the runtime operation.

Reply:
Yes we are, this checks whether we can promote and if not, we remove it from the list. Then we process this list in `promote_candidates`, which can only fail to promote without ICEing in case of nested promotion candidates, where promoting one candidate eliminated the other candidate because the other candidate is part of the already promoted one.

Reply:
I am entirely confused now.

Going back a step -- did you want to say above that my proposed comment was incorrect, and suggest a better one? I don't understand what is is you are trying to say here.

Reply:
Yes, I believe your comment is incorrect

```
// If we use explicit validation, we expect promotion to succeed.
// If an explicit promotion candidate were to get "unpromoted" here,
// that means we keep an operation that we expect to happen at compile-time
// at runtime. In case of repeat elements, this would just cause an error
// later in compilation (due to multiple moves out of a `!Copy` value), but
// for `rustc_arg_required_const` function arguments, where a constant is
// required, ending up with a runtime value would likely ICE the compiler
// during codegen, or silently cause undocumented and undesired LLVM features
// to be used.
```

Reply:
@RalfJung  What do you think about the above comment?

Reply:
I think we are having different views of the same thing here.

Re: your comment, I am not worried about the case where, if promotion analysis here fails, we ICE the compiler later. I am worried about the case where we do *not* ICE later. Because that means that, no matter whether promotion succeeds or fails, we produce a program, *but the programs may do different things* because we use the less conservative, "explicit promotion" rules. *That* is what I think this comment should explain.

Reply:
(This is the `delay_span_bug` that I meant in my other comment, that we should add nevertheless.)